ocs/cluster.py

```diff
@@ -0,0 +1,351 @@
+        """
+        Get accurate info on current state of pods
+        """
+        self._ceph_pods = pod.get_all_pods(self._namespace)
+        self.mons = pod.get_mon_pods(self.mon_selector, self.namespace)
+        self.mdss = pod.get_mds_pods(self.mds_selector, self.namespace)
+        self.mgrs = pod.get_mgr_pods(self.mgr_selector, self.namespace)
+        self.osds = pod.get_osd_pods(self.osd_selector, self.namespace)
+        self.toolbox = pod.get_ceph_tools_pod()
+
+        # set port attrib on mon pods
+        self.mons = list(map(self.set_port, self.mons))
+        self.cluster.reload()
+        if self.cephfs_config:
+            self.cephfs.reload()
+
+        self.mon_count = len(self.mons)
+        self.mds_count = len(self.mdss)
+        self.mgr_count = len(self.mgrs)
+        self.osd_count = len(self.osds)
+
+    @staticmethod
+    def set_port(pod):
+        """
+        Set port attribute on pod.
+        port attribute for mon is required for secrets and this attrib
+        is not a member for original pod class.
+
+        Args:
+            pod(Pod): Pod object without 'port' attribute
+
+        Returns:
+            pod(Pod): A modified pod object with 'port' attribute set
+        """
+        l1 = pod.pod_data.get('spec').get('containers')
+        l2 = l1[0]['ports'][0]['containerPort']
+        # Dynamically added attribute 'port'
+        pod.port = l2
+        logging.info(f"port={pod.port}")
+        return pod
+
+    def is_health_ok(self):
+        """
+        Returns:
+            bool: True if "HEALTH_OK" else False
+        """
+        self.cluster.reload()
+        return self.cluster.data['status']['ceph']['health'] == "HEALTH_OK"
+
+    def cluster_health_check(self, timeout=0):
+        """
+        Check overall cluster health.
+        Relying on health reported by CephCluster.get()
+
+        Args:
+            timeout (int): in seconds. By default timeout value will be scaled
+                based on number of ceph pods in the cluster. This is just a
+                crude number. Its been observed that as the number of pods
+                increases it takes more time for cluster's HEALTH_OK.
+
+        Returns:
+            bool: True if "HEALTH_OK"  else False
+
+        Raises:
+            CephHealthException: if cluster is not healthy
+        """
+        timeout = 10 * len(self.pods)
+        sample = TimeoutSampler(
+            timeout=timeout, sleep=3, func=self.is_health_ok
+        )
+
+        if not sample.wait_for_func_status(result=True):
+            raise exceptions.CephHealthException("Cluster health is NOT OK")
+        # This way of checking health of different cluster entities and
+        # raising only CephHealthException is not elegant.
+        # TODO: add an attribute in CephHealthException, called "reason"
+        # which should tell because of which exact cluster entity health
+        # is not ok ?
+        expected_mon_count = self.mon_count
```

Comment:
@shylesh 1 question. are we comparing the actual  mon count against the "count" specified in cluster.yaml or  **count** so as to confirm if required number of mons are UP ? 
```
oc get cephcluster rook-ceph -n openshift-storage -o yaml 
```
Ignore if its already the case.

Reply:
@nehaberry ,
Always self.mon_count tries to reflect number of mon pods in the cluster. Initially when we create CephCluster() object we do a scan_cluster which populates mon_count using a OCS.get() on cephcluster. Then further entry points to update mon count will be through mon_change_count(absolute_count)  and we will compare new count against the number of mon pods found using scan_cluster which again uses OCS.get(). So Its always from current cluster status.

I hope it answers your question.